### ISSUE TITLE
Documentation for "DeepSpeech setup" needs to be updated

### ISSUE THREAD

**Issue by TSRBerry** [NaomiProject/naomi-docs#20]
As the page itself already says, the documentation for DeepSpeech needs to be updated. I am currently experimenting with it, so I might be able to issue a PR for it. If my experiment works out, I will share the new steps needed in order to compile DeepSpeech correctly.

**Comment by AustinCasteel** [comment]
Thank you for the contribution. We have been busy with the plugin integrations so somethings have been neglected such as this, so any and all help if very much appreciated. With that being said, do note which version of Naomi you are using and adjust the corresponding docs (Unless you were planning on doing both stable & dev). That is something a lot of people do not realize when making documentation changes.

**Comment by TSRBerry** [comment]
I used the Naobian Image so I think I am using the stable version right?
So I should create a PR for the stable branch?

**Comment by AustinCasteel** [comment]
Its good to actually hear from someone that is using the image. Actually Naobian was built off of the dev branch in the state it was in on April 15th of this year. I am working on a new setup script that will be in the new Naobian build that will setup Naomi based on version selection as well as keep things up to date if the user so chooses, thus keeping docs concurrent to the actual build but it is still in the works.

Since the Naobian release there has been a good amount of changes to the dev branch. So my suggestion would be to save your profile somewhere safe and start over from the current state of dev to eliminate any issues or differences that may arise. For right now that can be done by downloading the dev source from the repo and setting up Naomi from it, or on the first of August I will be putting together another Milestone build and the websites manual download instructions will be updated as such to give a new baseline.

**Comment by TSRBerry** [comment]
Sounds good! Then I will test my solution on the current dev branch and will, depending on the result, write about it again. Or should I create a PR immediately with my changes to the commands?

**Comment by AustinCasteel** [comment]
If there are any issues/questions/concerns with the current state of dev you can write about it here in the issue, but if everything works out, then you can go ahead with the PR to the dev branch!

**Comment by TSRBerry** [comment]
The documentation seems to be for every Linux platform, right? I am just testing DeepSpeech on my Raspberry Pi so I am not sure if my settings will work for other distros/platforms as well.
So it could be the case that we need a lot more testing before I can create a PR to update the docu. Or will it automatically work for other platforms as well?

**Comment by AustinCasteel** [comment]
In theory Naomi will work on any linux distro as long as you have python. With that being said, there are things that in the code that are specifically for rpi as well as the documentation for some of the plugins just work for rpi due to the different nuances of each system (you can still use the plugins but the process of setting them up are different based on platform). Currently we are just focused on the rpi until we have a good baseline to branch out to other platforms officially.

Naturally the docs will have to work with the rpi3, but do not need to be specifically just for it. Take PocketSphinx for example, the instructions are for Raspbian & rpi as such. But they should still work with Debian based distros like Ubuntu, Mint, etc with little to no changes. So for the most part as long as it works on the rpi then things should be all good. Now if there is something that comes up, then there is someone else that we can bring on that has more knowledge in this area. @aaronchantrill

**Comment by TSRBerry** [comment]
Thanks!
It seems like you don't need to build DeepSpeech yourself in order to run it. I need to test it on a fresh installation, but I am pretty sure it should work by running:
`pip3 install DeepSpeech`
This installed the latest stable DeepSpeech version (v0.4.1) for me and it seems to work fine. You need to download the model files of course, but maybe we don't even need to build it ourselves anymore.
I spent the last 3 days building DeepSpeech on my Pi, so I could still update the documentation on how to build it yourself, but it might not be necessary. What do you think about that?

**Comment by TSRBerry** [comment]
I can confirm now:
After running the setup script from the latest Naomi-dev version, I was able to install DeepSpeech by using `sudo pip3 install DeepSpeech`.
After the installation finished all you need to do is downloading a model from: https://github.com/mozilla/DeepSpeech/releases/tag/v0.4.1 and extract it.

**Comment by AustinCasteel** [comment]
This actually jogged my memory, about a month ago I stopped working on the setup script since I was moving and at that time there was a new way to install DeepSpeech without building it. `sudo pip3 install DeepSpeech` But the concern that was raised at the time was that it really needs to be trained before it will do anything useful. I have not personally messed with it but that was where it was left before moving on to other things.

So then my question would be are you able to install DeepSpeech, extract the model, and get Naomi running with it? Or is there still training evolved?

**Comment by TSRBerry** [comment]
I downloaded the trained model from the DeepSpeech release and got Naomi working with it. I had some problems with the VAD but not with DeepSpeech. I didn't train the model myself it just worked.
(I am not to sure if that answers your question completely.)

**Comment by AustinCasteel** [comment]
Ok, I was not sure about the specifics of it. What was the VAD issue you had?

**Comment by TSRBerry** [comment]
I am not sure if I configured it the right way, but Naomi would just stop recording after a second regardless of the config in profile.yml. But if it records for a longer time it won't stop recording until I say the key word again and afterwards it won't react to anything.

**Comment by aaronchantrill** [comment]
@jeliebig The default VAD usually works well for me, but there are things that cause the behavior you are experiencing. The first/easiest thing you might want to try is switching over to WebRTCVAD: https://projectnaomi.com/dev/docs/configuration/vad.html#webrtc-vad
The default VAD is based on collecting audio volume levels and then setting a threshold at 1.5* standard deviation of the volume levels. This is because different sound cards report different volume levels, and different environments have different baseline audio level (such as if there is a fan running in the room). Sometimes it can act a bit wonky when it first starts up, especially if your audio contains pops, but it usually calms down after 20 or 30 seconds. If you run Naomi with the --debug flag, you will be able to see the SNR (which is basically the noise ratio), the threshold, mean, etc. This should give you a better idea of what is going on.

**Comment by aaronchantrill** [comment]
I would also recommend adding the following settings to your ~/.naomi/configs/profile.yml:
```
print_transcript: true
audiolog:
  save_audio: true
```
The first allows you to see what Naomi thinks it hears in real time. The second captures everything Naomi thinks is important enough to listen to into wav files in the ~/.naomi/audiolog directory. There is also a sqlite3 database located at ~/.naomi/audiolog/audiolog.db. The easiest way to interact with this is to run ./NaomiSTTTrainer.py from Naomi's base directory (wherever git put the files). This can be a big help in identifying issues with your microphone.

**Comment by aaronchantrill** [comment]
@jeliebig Thanks for looking at the DeepSpeech documentation. I think the old documentation stopped working about a year ago, but I left it up in case someone wanted to install from source. But certainly at this point all they have to do is run the pip install which is way easier.

**Comment by AustinCasteel** [comment]
Along those lines, if it is just a couple lines then they can be added to the stt page just like the google cloud info and we can remove the deepspeech dedicated page.

**Comment by TSRBerry** [comment]
@aaronchantrill No problem, I thought my experiment could be useful to someone that might want to build DeepSpeech themself, that's why I opened this issue.
I also found the debug flag and was able to kinda solve the issue, but I wasn't able to use WebRTC.
Even if I add the 'input_samplerate' line to my audio settings in 'profile.yml' WebRTC still throws an exception. It seems like WebRTC doesn't process the captured frames the right way because the exception says that there was an error with input rate. I tried all the accepted numbers in the config file but it wouldn't work. So I gave up on WebRTC and switched back to default VAD and reconfigured the threshold.
@AustinCasteel Yup, here are the lines I needed to install DeepSpeech on the latest dev version:
`sudo pip3 install DeepSpeech`
`wget https://github.com/mozilla/DeepSpeech/releases/download/v0.4.1/deepspeech-0.4.1-models.tar.gz`
`tar xzvf deepspeech-0.4.1-models.tar.gz`
And then I added these lines to my profile.yml:

```
active_stt:
  engine: deepspeech-stt
deepspeech:
  model: '/home/pi/models/output_graph.pb'
  alphabet: '/home/pi/models/alphabet.txt'
  language_model: '/home/pi/models/lm.binary'
  trie: '/home/pi/models/trie'
```

**Comment by aaronchantrill** [comment]
Interesting. What sound card are you using?

**Comment by aaronchantrill** [comment]
Actually, I noticed that I'm having the same problem with WebRTCVAD when using the default Intel 82801AA-ICH card created by VirtualBox to act as a passthrough to my host computer's default audio card. My chunksize is set to 960 bytes or 480 samples, but the chunks actually arriving at WebRTC are 682 bytes long, or 341 16 bit samples. I'm digging into it now.

**Comment by TSRBerry** [comment]
I am using my Logitech G430 USB Headset

Thank you for looking into that!

**Comment by aaronchantrill** [comment]
I played around a bunch with this last night. With regard to the default device on VirtualBox under alsa, I'm getting this error when I select plughw-card-i82801aaich-dev-0: "alsaaudio.ALSAAudioError: Invalid argument [plughw:CARD=I82801AAICH,DEV=0]". Now "aplay -Dplughw:CARD=I82801AAICH,DEV=0 ~/test.wav" works perfectly well, so I'm not sure what the "invalid argument" is, and I haven't dug into the cython to see if I can figure out what is actually generating this error.

I was able to test the card using "sysdefault-card-i82801aaich" ("sysdefault:CARD=I82801AAICH") but it was always returning a chunk of size 682 bytes no matter what I set period size to. I tried cutting this down and only using the first or last 640 bytes (320 16 bit samples), and that appears to have worked. I'm putting in a pull request for that.

As far as the snr_vad goes, I was testing it last night while trying not to wake up my son in the next room, so kind of talking under my breath and trying not to project, and I had the problems you described. I changed the line "threshold = mean + (stddev * 1.5)" to "threshold = mean + (stddev * 1)" and that seemed to fix it. That is also working for me today, while I can speak freely, so I'm going to change that to the default, but I seem to be missing something. In both cases, the stddev was about 7, so it doesn't appear to be responding to a less projected voice with a lower stddev and thus threshold, which is what I was hoping. In the meantime, I can make the multiplier something that can be set in profile.yml, something like "aggressiveness" in webrtcvad, but opposite. tolerance, perhaps?

**Comment by aaronchantrill** [comment]
I just installed Rasbian Buster and ... deepspeech has not yet been precompiled for python 3.7. So I guess I'll be installing from scratch again this weekend.

**Comment by TSRBerry** [comment]
Jup, the latest version of DeepSpeech (https://pypi.org/project/deepspeech/) was just precompiled for Python 3.6 and some older versions.

**Comment by aaronchantrill** [comment]
I tried installing Naomi on Buster using naomi-seup.sh and choosing option 2 "Download and install a local copy of Python for Naomi" which downloaded and installed Python 3.5. From there I was able to install virtualenvwrapper with this custom python3.5 as the default.  That allowed pip to download and install deepspeech. Unfortunately now I am getting an error about "no module named _impl" (this cython module is part of deepspeech itself). I'm reformatting and starting over from scratch to see if I can get this to work.

**Comment by aaronchantrill** [comment]
Hey @jeliebig I think this is a good candidate for a Hacktoberfest issue. Are you still working on this? If not, I'd like to open it up.

**Comment by TSRBerry** [comment]
Thanks for mentioning me!
I actually kinda stopped after the issues I had with the speed of DeepSpeech but I do want to update the documentation with my findings if that is still okay.
Should I include the steps to compile it yourself and the precompiled version? Or should I just include the short pip install with the lastest DeepSpeech version?

**Comment by aaronchantrill** [comment]
@jeliebig I know deepspeech is a lot slower than pocketsphinx, but what issues did you have exactly?

I'm extremely interested in going deeper with both Deepspeech and Kaldi. My issue with DeepSpeech was that on the Raspberry Pi, the whole acoustic dictionary could not be loaded into memory at once, which was required at the time. It worked without that, but only returned the raw phonemes and was unable to translate to full words. Which is actually okay for me, because I would like to do a soundex type comparison at the phoneme level when scanning for intent so you don't have to worry about whether the STT processor heard "two", "to" or "too" when the user says "add eggs to my shopping list".

I feel like trimming down the dictionary a lot could be a big help and still allow the user to say generic terms. I'm also looking at using pocketsphinx to parse the audio into clips and then just send the bit of audio that represents a variable in the intent to pocketsphinx. For instance, you could have an intent like "add {item} to shopping list" which would match "add which to shopping list" or whatever odd match pocketsphinx returns. Using the pocketsphinx python tools, we can see that "add" ends on frame 115 and "to my shopping list" starts on frame 153, so we clip frames 116 to 152 and send that to pocketsphinx for a deeper analysis, which will hopefully return the correct term.

As far as getting deepspeech installed from source, I really don't think it's necessary unless you are planning to develop on deepspeech yourself, in which case you probably have the experience required to get it installed, and you really need to be running it on a desktop with a beefy GPU. The deepspeech requirements and dependencies change much too quickly for us to keep track of, so unless you have a good rundown of exactly how to install from source, I'd say that just the Pypi install plus explaining how to download and configure the correct model files for the version you end up would be sufficient. If you can find instructions for manually installing deepspeech on the mozilla website, you might want to throw in a link to that in case anyone does want to work on development, or at least analyzing the source. I think this would also be a good place to plug the "Common Voice" project and explain that contributing to that project can improve deepspeech recognition.

**Comment by TSRBerry** [comment]
Yeah, I was having exactly the issue that you described and although I knew about it, it felt too slow for me to actually use it with my project. So it's nothing new and I was just unhappy with it after some time.

Thank you so much for helping me with my PR!
I don't have in-depth knowledge about pocketsphinx or any STT related stuff, but the context idea sounds great from my understanding and could help to get better results more often. The "coding idea" behind this would be to search the recognized audio for "add {item} to shopping list" and trigger a context related event? This also makes more room for plugins and I really think that this would be awesome!

**Comment by AustinCasteel** [comment]
Fixed issue with [PR #24](https://github.com/NaomiProject/naomi-docs/pull/24). If issues come up again the issue can be reopened or a new one can be created. @jeliebig Thanks for the PR!
